In [2]:
import mxnet as mx

[Mxnet 'Getting Started' guide](http://mxnet.readthedocs.io/en/latest/get_started/index.html)

### Part one

In [2]:
train = mx.io.MNISTIter(
    image      = "mnist/train-images-idx3-ubyte",
    label      = "mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
val   = mx.io.MNISTIter(...)

MXNetError: [09:03:57] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "mnist/train-images-idx3-ubyte"

Downloaded mnist's 'train-images-idx3-ubyte' manually into mxnet/mnist/train-images-idx3-ubyte

In [4]:
train = mx.io.MNISTIter(
    image      = "mnist/train-images-idx3-ubyte",
    label      = "mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
val   = mx.io.MNISTIter(...)

MXNetError: [09:13:23] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "mnist/train-images-idx3-ubyte"

Downloaded all three remaining files

In [5]:
train = mx.io.MNISTIter(
    image      = "mnist/train-images-idx3-ubyte",
    label      = "mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
val   = mx.io.MNISTIter(...)

MXNetError: [09:15:13] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "mnist/train-images-idx3-ubyte"

In [4]:
train = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

Adding the whole path worked with the first part but this next part has now a problem

In [7]:
val = mx.io.MNISTIter(...)

MXNetError: [09:47:44] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "./train-images-idx3-ubyte"

In [8]:
val   = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

I supose I should do that, as the val is a validation set.

### Part two

In [9]:
data = mx.symbol.Variable('data')
fc1  = mx.symbol.FullyConnected(data = data, num_hidden=128)
act1 = mx.symbol.Activation(data = fc1, act_type="relu")
fc2  = mx.symbol.FullyConnected(data = act1, num_hidden = 64)
act2 = mx.symbol.Activation(data = fc2, act_type="relu")
fc3  = mx.symbol.FullyConnected(data = act2, num_hidden=10)
mlp  = mx.symbol.SoftmaxOutput(data = fc3, name = 'softmax')

### Part three

In [10]:
model = mx.model.FeedForward(
    symbol = mlp,
    num_epoch = 20,
    learning_rate = .1)
model.fit(X = train, eval_data = val)

### Part four

In [23]:
test = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
model.predict(X = test)

array([[  3.33800598e-10,   2.53869571e-13,   6.84752477e-10, ...,
          3.31243852e-11,   3.54347407e-07,   2.39830683e-07],
       [  1.78424549e-11,   8.90376894e-15,   7.65590253e-16, ...,
          1.75893402e-18,   2.36171360e-07,   1.27299768e-06],
       [  8.31436135e-12,   3.35610175e-13,   6.77721488e-08, ...,
          7.25360476e-15,   9.99999881e-01,   1.01463048e-11],
       ..., 
       [  3.51093621e-09,   1.09971005e-07,   2.04552268e-08, ...,
          3.52970318e-07,   1.31745655e-05,   4.23762569e-04],
       [  1.13982500e-12,   3.06890337e-13,   5.46858565e-13, ...,
          1.50954704e-09,   5.87727145e-10,   1.39010785e-06],
       [  4.17790761e-05,   3.65532798e-10,   1.04760488e-06, ...,
          9.78314340e-01,   3.14737804e-06,   2.15601176e-02]], dtype=float32)

### Trying to understand what that means

In [26]:
test = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
prediction = model.predict(X = test)

In [27]:
prediction.shape

(59904, 10)

So the output is 59904 results, one for each image in the test dataset. For every row (result) there are 10 probabilities, one for each digit

In [31]:
pred_last_ten = prediction [-10:]

Taking the last 10 predictions

In [39]:
for i in range(pred_last_ten.shape[0]):
    result = 'result {} was inconclusive'.format(i)
    for j in range(pred_last_ten.shape[1]):
        if pred_last_ten[i][j] > 0.9:
            result = 'result {} prediction is {}'.format(i, j)
    print(result)


result 0 prediction is 4
result 1 prediction is 3
result 2 prediction is 6
result 3 prediction is 2
result 4 prediction is 4
result 5 prediction is 4
result 6 was inconclusive
result 7 prediction is 4
result 8 prediction is 4
result 9 prediction is 7


Were there really those numbers?